### PR TITLE
Add initial code

### PR DESCRIPTION
**Pull request by Tasssadar**
Are any other API methods necessary? Calibration, stuff like that?

Sending values to lorris:

```cpp
    LineSensor line;
    ESP_ERROR_CHECK(line.install());

    uint8_t buff[4 + (LineSensor::CHANNELS*2) + 4] = { 0x55, 0x55, 0, 1 };
    buff[2] = sizeof(buff) - 4;

    while(1) {
        ESP_ERROR_CHECK(line.read((uint16_t*)(buff+4)));

        const float pos = line.readLine(false, 0xFF, 0.1f, 0.2f);
        *((float*)(buff + 4 + (LineSensor::CHANNELS*2))) = pos;

        Serial.write((uint8_t*)buff, sizeof(buff));
        vTaskDelay(20);
    }
```

https://cdn-shop.adafruit.com/datasheets/MCP3008.pdf

### CODE REVIEW

File: src/linesensor.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * \brief The LineSensor device - one MCP3008 chip.
+ *        It is not thread-safe, you have to make sure the methods are called
+ *        from one thread at a time only.
+ *        The @{install} method has to be called before you can use any other methods.
+ */
+class LineSensor {
+public:
+    static constexpr int CHANNELS = 8; //!< Amount of channels on the chip
```
Comment (JarekParal) [inline src/linesensor.h]: If is this library universal library for MCP3008, the number of channels should be a parameter of the config.

File: src/linesensor.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+     */
+    esp_err_t uninstall();
+
+    /**
+     * \brief Read values from the chip. Returns values in range <0; @{MAX_VAL}>.
+     *
+     * \param results the results will be APPENDED to this vector.
+     *        It will be unchanged unless the ESP_OK result is returned (except possibly its capacity).
+     *        Between 0 and @{CHANNELS} values are appended, depending on the @{channels_mask} param.
+     * \param differential return differential readings, as specified in the MCP3008 datasheet.
```
Comment (JarekParal) [inline src/linesensor.h]: I don't think that is this parameter good in the LineSensor class. It doesn't make sense to use it on a line sensor board.

File: src/linesensor.h (hunk 1 above)
Comment (JarekParal) [inline src/linesensor.h]: If is this library universal library for MCP3008, the number of channels should be a parameter of the config.

File: src/linesensor.h (hunk 2 above)
Comment (JarekParal) [inline src/linesensor.h]: I don't think that is this parameter good in the LineSensor class. It doesn't make sense to use it on a line sensor board.

**Review by JarekParal — COMMENTED** [review]
It would be nice to have some function to calibrate the line sensor. But how to implement that without the hardware control (move the robot around the line...)?

File: src/linesensor.h (hunk 1 above)
Comment (Tasssadar) [reply src/linesensor.h]: Well MCP3008 is always going to have 8 channels. We could move the `channels_mask` param from all the functions below and make it a one-time configuration value though.

File: src/linesensor.h (hunk 2 above)
Comment (Tasssadar) [reply src/linesensor.h]: It's supposed to be generic for the MCP, you can use this library even for other things - then you only use the read method and nothing else.

**Comment by Tasssadar** [comment]
> It would be nice to have some function to calibrate the line sensor. But how to implement that without the hardware control (move the robot around the line...)?

I didn't do that yet because it is hard to test when just holding the sensors in hand, need something to hold the whole board at constant height.
The "moving the robot" part will be up to the user's code, the library will just record the values when told to and will keep the calibration values.

File: src/linesensor.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    for(int i = 0; i < CHANNELS; ++i) {
+        if(((1 << i) & channels_mask) == 0)
+            continue;
+
+        auto &t = transactions[i];
+        t.user = (void*)requested;
+        t.flags = SPI_TRANS_USE_RXDATA | SPI_TRANS_USE_TXDATA;
+        t.length = 3 * 8;
+        t.tx_data[0] = 1;
+        t.tx_data[1] = (differential << 7) | ((i & 0x07) << 4);
```
Comment (Tasssadar) [inline src/linesensor.cpp]: Note to self: bug, `(0 << 7)` means differential and `(1 << 7)` means single-ended, it's the other way around here.

File: src/linesensor.h (hunk 1 above)
Comment (Tasssadar) [reply src/linesensor.h]: Moved the mask to the config.

**Comment by Tasssadar** [comment]
I think this is enough for the initial implementation, I'll wait for your review and then merge it.

**Comment by JarekParal** [comment]
@yaqwsx Could you please look at this PR if is everything clear and good from your point of view?

**Comment by yaqwsx** [comment]
The review was made personally.

**Comment by Tasssadar** [comment]
@yaqwsx Did the Driver/Sensor split, pls take a look.

**Comment by yaqwsx** [comment]
Seems OK to me.

**Pull request merged**